### ISSUE TITLE
Improve pie chart / circle performance

### ISSUE THREAD

**Issue by aappling-usgs** [USGS-VIZLAB/water-use-15#97]
Make it faster to load the page, and transition into and out of pie charts and circles, 

Relates to #71 (basic pie charts), #72 (spiffy pie chart transitions), and #62 (load time performance), and messing with transitions and loadable data are within scope (despite overlapping with those other issues) to the extent that they affect performance.

**Comment by aappling-usgs** [comment]
Lindsay and Alison have already been working on this for several hours today (https://github.com/lindsaycarr/water-use-15/commit/eafe0893aa0c10940ded20e90ad1dc54bfedf489 through https://github.com/lindsaycarr/water-use-15/commit/0269bf314869fabf7e1d3a07213ea4adf8b1c188). We've made it so the pie groups and pie slices only get created once per session, moved the pie chart data preparation into build_map so that also only happens once now, and temporarily turned off transitions to be able to view the changes faster. As a baseline from there (I'm hoping to find additional opportunities for improvement), here are 
* load times to `view=USA&category=total` (milliseconds):

total | scripts (all the yellow) | rendering (all the purple) | painting (all the green) | 'load' handler took __ms
-- | -- | -- | -- | -- 
1507 | 1229 | 77 | 5.5 | 873
2745| 1351 | 82 | 5.3 | 949
1625 | 1313 | 79 | 5.3 | 940
1551 | 1246 | 79 | 5.6 | 873
1556 | 1254 | 80 | 5.3 | 900

* load times to `view=USA&category=piechart` (milliseconds) (with `Error: <circle> attribute r: Expected length, "NaN".`):

total | scripts (yellow) | rendering (purple) | painting (green) | 'load' handler
-- | -- | -- | -- | -- 
2967 | 1749 | 80 | 3.4 | 1201
2166 | 1782 | 77 | 3.2 | 1372
3204 | 1776 | 75 | 3.3 | 1390
2096 | 1728 | 81 | 3.9 | 1340
2137 | 1776 | 85 | 2.5 | 1404

* times for first transition from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler
-- | -- | -- | -- | -- 
?? | 473 | 302 | 26.6 | 428
?? | 520 | 310 | 27.5 | 464
?? | 536 | 307 | 28.4 | 497
?? | 517 | 303 | 27.5 | 474
?? | 561 | 301 | 26.7 | 534

* times for subsequent transitions from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler took __ms
-- | -- | -- | -- | -- 
?? | 227 | 275 | 25.7 | 224
?? | 227 | 293 | 28.2 | 225
?? | 213 | 284 | 26.5 | 213
?? | 217 | 275 | 24.7 | 215
?? | 296 | 272 | 25.6 | 195

* times for mouseouts back to from Pie to Total view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseout' handler took __ms
-- | -- | -- | -- | -- 
?? | 167 | 264 | 8.2 | 162
?? | 266 | 337 | 8.5 | 145
?? | 153 | 270 | 8.4 | 150
?? | 167 | 273 | 8.0 | 165
?? | 243 | 273 | 8.2 | 152

**Comment by aappling-usgs** [comment]
After updating so that pie charts are scaled with `transform(scale())` rather than having county-specific radii in the arcs:

* load times to `view=USA&category=total` (milliseconds):

total | scripts (all the yellow) | rendering (all the purple) | painting (all the green) | 'load' handler took __ms
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* load times to `view=USA&category=piechart` (milliseconds) (with `Error: <circle> attribute r: Expected length, "NaN".`):

total | scripts (yellow) | rendering (purple) | painting (green) | 'load' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* times for first transition from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* times for subsequent transitions from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* times for mouseouts back to from Pie to Total view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseout' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | --

**Comment by aappling-usgs** [comment]
@jread-usgs , laura mentioned in standup that you'd found some nifty way of representing everything as one giant path (or something) in gages through the ages, which was the deal-maker for getting that vizzy to be performant for a lot of gages. do you remember what you did and/or where the relevant issues or PRs might be?

**Comment by aappling-usgs** [comment]
[not finished but my computer is going to reboot]

Now I've reverted to county-specific radii, have pushed more of the data preparation work into `pieData()`, and have attempted to tighten up here and there. Not expecting big improvements, but also don't see other easy ways to speed things up, so here goes:

* load times to `view=USA&category=total` (milliseconds):

total | scripts (all the yellow) | rendering (all the purple) | painting (all the green) | 'load' handler took __ms
-- | -- | -- | -- | --
3000 | 1241 | 84 | 27 | 818
3066 | 1242 | 83 | 17 | 822
2999 | 1252 | 82 | 19 | 837

* load times to `view=USA&category=piechart` (milliseconds) (still with `Error: <circle> attribute r: Expected length, "NaN".`, and no rendering of the pies):

total | scripts (yellow) | rendering (purple) | painting (green) | 'load' handler
-- | -- | -- | -- | -- 
3533 | 1675 | 75 | 5.2 | 1261
3615 | 1835 | 77 | 4.3 | 1318
3477 | 1696 | 86 | 4.6 | 1270
2064 | 1675 | 77 | 2.5 | 1295
3296 | 1761 | 78 | 4.7 | 1352

* times for first transition from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* times for subsequent transitions from Total to Pie view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseover' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 

* times for mouseouts back to from Pie to Total view

total | scripts (yellow) | rendering (purple) | painting (green) | 'mouseout' handler
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | -- 
-- | -- | -- | -- | --

**Comment by aappling-usgs** [comment]
ideas still to try:
* reduce the overall number of polygons (and enable nifty wiper transitions) by implementing each circle as a pie slice, so there are 5 rather than 6 polygons per county (plus 1 group)
* get rid of the transparency
* don't even plot the tiny circles
* don't even plot the tiny pie slices
* plot tiny circles as a single path containing rounded points, rather than as a bunch of tiny circles
* hide or alpha-out rather than shrink the circles/pies when not using them

**Comment by jread-usgs** [comment]
One potential gotcha on drawing a circle with a path in svg - you can't draw the full circle with the arc. 

[The simple solution](https://stackoverflow.com/questions/5737975/circle-drawing-with-svgs-arc-path) is to draw 99.9% of the circle and then close w/ `z` ("close" polygon in svg-path-speak)

**Comment by jread-usgs** [comment]
and maybe also in order to get rid of all possible slow-down sources, remove the tool-tip function for the time being?

**Comment by jread-usgs** [comment]
Can the circles be pre-sorted (like in the geojson)?

**Comment by aappling-usgs** [comment]
Lindsay suggested that, too, but then we thought more about it and couldn't find a cheap way to make them pre-sorted for all categories. So now Lindsay is trying out a way to not have to sort them at all (by moving tooltip behavior to be county-centric rather than dot centric). If it works technically, this change will involve a check from our design lead :wink: , but Marty and Lindsay and I agreed that it might be sufficiently intuitive and a good move if it improves performance a lot. (Sorting is currently responsible for about half of the Scripting time in each switch among categories, so it probably will.)

**Comment by aappling-usgs** [comment]
Options for showing/hiding items:
 1
   .style("display","none");
   .style("display", null);
 2
   .attr("hidden",true); //to hide
   .attr("hidden",null); //to show
 3
   .style("opacity",0); //to hide
   .style("opacity",1); //to show

**Comment by aappling-usgs** [comment]
jotted notes on css animations & scaling:

```
    .style("transform", function(d) {
      return "scale3d( " + ... + ")";
    });
    // style attr for scaling: ease-in, ease-out
```
and from climate fish viz https://owi.usgs.gov/vizlab/climate-change-walleye-bass/wallyTrendsFig-desktop.svg
```
@keyframes move-wally-7{
  0% {transform: translate(11px, -7px)}
  33% {transform: translate(-14px, 4px)}
  67% {transform: translate(15px, -1px)}
  100% {transform: translate(11px, -7px)}
}
#walleye-7 {animation: move-wally-7 7s ease-in-out infinite;} 
```

**Comment by aappling-usgs** [comment]
Not working, no. I'm still just working on getting a basic translate and scale to work with `style()`. I've tried splitting up the translate and scale bits into different calls with a tremendous lack of success, so now I'm back to defining them together, but even with that i can't seem to get them to show up in the inline css. the internet seems to always use `.attr("transform", function(d) ...)` rather than `.style(...` for transforms. I can use `.style()` for other things like fill and stroke-width, no problem, but can't seem to figure out transform.

leaves the inline css empty:
```
    .style(function(d) {
      return {
        transform: "translate(" + d.coordinates.x + " " + d.coordinates.y + ")"+
                   "scale(0 " + scaleCircles(d.properties[[newCat]]) + ")"
      };
    });
```

modifies the inline css:
```
   .attr("transform", function(d) {  
      return "translate(" + d.coordinates.x + " " + d.coordinates.y + ")"+
             "scale(0 " + scaleCircles(d.properties[[newCat]]) + ")";
    });
```

**Comment by jread-usgs** [comment]
I am surprised that the second example modifies the inline css and not the transform svg attribute. I would think you would do this to modify the inline css (`style`):
```
.attr("style", function(d) {  
      return "transform: translate(" + d.coordinates.x + " " + d.coordinates.y + ") scale(0," + scaleCircles(d.properties[[newCat]]) + ");";
    });
```
I think your example does this:
`<circle ... transform = "translate(1,2)scale(0 4);"/>`
but I think you want
`<circle ... style = "transform: translate(1,2) scale(0,4);"/>`
for the transform to by a CSS transformation instead of an svg transformation.

**Comment by jread-usgs** [comment]
Geez, not much luck on my end. I couldn't get the different `style` things to play nice. 

I do have a simple svg here with what I think we are going for in terms of transition from one category to the next:
```xml
<?xml version="1.0"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="xMinYMin meet" viewBox="0 0 200 200">
	<style>
		
		@keyframes coin-flip {
		  0% {
		   transform: scale(1,1);
		   fill:red;
		  }
		  49% {
		   transform: scale(0.1,1.3);
		   fill:red;
		  }
		  50% {
		   transform: scale(0.1,1.3);
		   fill: green;
		  }
		  100% {
	   		transform: scale(1.6,1.6);
	   	 	fill: green;
		  }
		}
	#test:hover{ 
	  animation: coin-flip 1s;
	  animation-timing-function: linear;
	}
	</style>
	<g transform="translate(75,75)">
		<circle r="40" stroke="black" stroke-width="10" fill="red" id='test' vector-effect="non-scaling-stroke"/>
	</g>
</svg>
```
if you open that in a browser you can mouseover and see. I don't come up with a simple way to do this w/ css that didn't use keyframes. There are chained examples as has been noted, but what I don't see in the chaining in doing a scale, then something, then another scale. 
hmm...

**Comment by jread-usgs** [comment]
I see the issue w/ scale/translate in the same transform element. It does get wiped and needs to be updated to work. instead, you can create an outer group that holds the translate, and put the scale in an inner group. This is working example:
```xml
<?xml version="1.0"?>
<html prefix="og: http://ogp.me/ns#" lang="en">
<head>
	<script src="https://d3js.org/d3.v4.min.js" charset="utf-8"></script>
</head>
<body>	
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" preserveAspectRatio="xMinYMin meet" viewBox="0 0 5 5">
	<g transform="translate(2,2)">
		<g id = 'pie-group'>
			<circle r="1" stroke="red" stroke-width="10" fill="red" id='test' vector-effect="non-scaling-stroke"/>
			<g id='pie'>
				<path d='M 1 0 A 1 1 0 0 1 0.809017 0.587785 L 0 0'; fill="coral"/>
				<path d= 'M 0.809017 0.587785 A 1 1 0 1 1 -1.83697e-16 -1 L 0 0' fill='green'/>
				<path d='M -1.83697e-16 -1 A 1 1 0 0 1 0.951057 -0.309017 L 0 0' fill='rgb(0, 171, 107)'/>	
			<g/>
		<g/>
	</g>
</svg>
	<script>

		
	  d3.selectAll('#pie-group')
		.transition()
		  .ease(d3.easeLinear)  
		  .duration(500).delay(0)          
		  .attr('transform', 'scale(.1,1.3)')
        .transition()
          .duration(500)
          .attr('transform', 'scale(1.6,1.6)')
	  
	  d3.selectAll('circle')
        .transition()           
          .duration(0).delay(500)
   		  .style('display','none');
	
 	  d3.selectAll('#pie')
 		.style('display', 'none')
		.transition()
          .duration(0).delay(500)
          .style('display', null)

	</script>
</body>
</html>
```

**Comment by jread-usgs** [comment]
- seems we can re-use something like the first block of this transition regardless of which categories are being moved to/from, since it only deals with the to/from scales
 - Transitioning to or from the pie will always use the same third block, with the exception of switching the position of `"none"` and `null` in the `.style()` calls
 - the "to" transition to a new tin category can all be done w/in the `.style()` too, with a function for what to change.

**Comment by lindsaycarr** [comment]
For me, time spent on #102 was 3 hrs.

**Comment by aappling-usgs** [comment]
I've spent 23 hours on this as of #109

**Comment by jread-usgs** [comment]
I have a few experiments (happy to share code to generate this, but I didn't want to check it into the repo). I wanted to isolate the challenge of figuring out what is slowing the transitions down. 

 - generate 2,000 circles w/ radius that has a lognormal distribution. 
 - represent the all of the circles as a single `<path>` (using the arcs) and represent the circles normally, as `<circle>`, all w/ radius of 1 and with a scale() transform. All in all, 4,000 circles on the page, w/ 2,000 represented by one path on the left and the same 2,000 represented by 2,000 circles on the right. 
 - create different transforms that change the radius and change the color. 

Findings: 

 - It seems that any transition that interpolates colors on a large number of different things using a css transition is going to struggle. If we instead use d3's transitions and put the fill on a different element than the element(s) that have their geometries changing, this works [bostock example](https://bl.ocks.org/mbostock/5348789). I used `defs` and transitioned the geometry on the element in `defs` and transitioned the style in the element in `use`.  
 - If we don't want to blend colors w/ a transition, using a transition that lets us to do a quick pivot to a new color seems to work well. The "coin flip" works and another transition that goes from the first radius to zero then second radius while changing the color/class at r = 0 works well too. I couldn't think of other ways to "blink" the color w/o it feeling jarring. 
 - D3's transition works really well at parsing a path's `d` attribute and interpolating, even if that path is _really_ long. I put all of the circles in a single path, and the transition was smooth. 
 - For the "coin flip", the circle and path representations of the scaled circles have different rending profiles:
path:
![image](https://user-images.githubusercontent.com/2349007/38169641-44d95556-3535-11e8-8820-74d487c49af1.png)
circle:
![image](https://user-images.githubusercontent.com/2349007/38169646-584b829e-3535-11e8-8743-befaf399fc6e.png)

Visually, they look pretty similar to me unless I really focus on the midpoint of the transition and the path is a bit smoother. If you throttle the CPU w/ chrome devtools, then the circle one starts to get really clunky. 

Here is what the two look like before transition:
![image](https://user-images.githubusercontent.com/2349007/38169669-b0785c1c-3535-11e8-9ec0-378cebb82b6d.png)
and after:
![image](https://user-images.githubusercontent.com/2349007/38169673-bf2fe522-3535-11e8-9b38-211803fd706c.png)

Using a path makes things harder for getting data on the map, but, like w/ gages through the ages, it likely helps w/ performance. Paths also let us keep a consistent stroke (but note this would be possible w/ circles too if we didn't use the coin-flip and instead took the radius to zero and then back up). Probably not enough of a reason to shift though. I will share the example html in our slack channel

**Comment by aappling-usgs** [comment]
Cool! So:
1) use a single path for all circles (one path per category)
2) use d3, not css after all, to do the transition
3) it was initially looking like it was important to separate the geometry and color transitions, but now you're finding that a single transition on both works well for circle-paths, is that right?

With those conclusions in mind, it sounds to me like color fades while the circles grow or shrink may perform just about as well as coin flips with a quick color transition in the middle. Is that true? And my initial reaction to coin flips (with the caveat that they were choppy in my implementation) is that the color fade and growing/shrinking actually looked a bit nicer. What's your preference now that you've seen both working smoothly?

**Comment by jread-usgs** [comment]
1. I think so, but I want to try one more test that uses the `r` instead of `transform` in the transition. I will add it to my experiment and report back. Should be simple, but I need to finish taxes first. 
2. definitely. d3 is smart w/ these things :+1: I'm sold.  
3. yes - w/o the pie charts, this is no longer necessary, and we can take advantage of a single transition for both geoms and style. This is simpler and a little smoother. I agree that the resize and re-color transition looks the best.

**Comment by aappling-usgs** [comment]
All sounds good! Heh, I just finished my taxes too.

**Comment by aappling-usgs** [comment]
So you're saying that `r` vs `transform` doesn't matter at all, and that only path vs circles matters? Or would r vs transform matter in some cases?

**Comment by jread-usgs** [comment]
@aappling-usgs r vs transform doesn't seem to matter for performance, although I would guess there is a slight advantage to `r` since D3's parser and interpolater has to do _slightly_ less work. The main difference between the two here is that `r` keeps a consistent stroke width (preferred) vs `transform='scale(x,y)'` scales the stroke, which is less than ideal. That is what I meant by "Now the path and circle implementation look identical to my eyes", but I should have been clear about that. The `path` implementation also keeps the stroke width the same, so that is what I was referring to.

**Comment by aappling-usgs** [comment]
ahhh, cool, thanks. great, `r` and `path` seems like a winning team

**Comment by wdwatkins** [comment]
Haven't read this whole thread, so sorry if this is already noted, but I noticed some things while messing with the analytics:

Looks like we reload a category whenever the button is hovered over, even if that category is already loaded.  Same with the mouseout.  There should be a check there to see if the category is already loaded.

And there seems to be mouse-in/mouse-out happening when you click on a category, not sure that should be happening.  I would think it is less snappy since it is loading the new circles twice.

**Comment by wdwatkins** [comment]
Might we also want to add a very small delay to deciding to load a new category, so just incidentally dragging the cursor over the buttons doesn't trigger category changes?  Although if the category changes get snappier this might not be an issue.

**Comment by lindsaycarr** [comment]
@wdwatkins that's a good idea. I think we should create a separate issue for it (@aappling-usgs do you agree?). Here's an example for click vs double click and I think that's pretty much what we would use for mouseover vs click. We should also handle mouseout differently if it's been clicked. No need to revert the category.

https://github.com/almende/vis/issues/203#issuecomment-48481501

**Comment by aappling-usgs** [comment]
OK, issue created. Thanks @wdwatkins and @lindsaycarr - great observations.

**Comment by aappling-usgs** [comment]
two major ideas have been spun off into their own issues:
* use r instead of transform or otherwise fix the circle boundary width on zoom (#96)
* switch back to the resize and re-color transition (#131)

what i think remains:
* try removing the tooltip function entirely, see if that speeds things up
* get rid of the transparency, see if that speeds things up (do this after #96, because might interact with circle vs path)
